browser/src/layer/tile/CanvasTileLayer.js

```diff
@@ -1527,26 +1480,37 @@ L.CanvasTileLayer = L.Layer.extend({
 			return this._tiles[key];
 		}
 		var tile = {
-			wid: 0,
 			coords: coords,
 			current: true, // is this currently visible
-			retain: false, // was it previously visible cf. _pruneTiles
 			canvas: null,  // canvas ready to render
 			rawDeltas: null, // deltas ready to decompress
-			loaded: false, // is the tile data present & valid
+			viewId: 0, // canonical view id
+			wireId: 0, // monotonic timestamp for optimizing fetch
+			invalidFrom: 0, // a wireId - for avoiding races on invalidation
 			lastRendered: new Date(),
 			hasContent: function() {
-				return tile.canvas || (tile.rawDeltas && tile.rawDeltas.length > 0);
+				return this.canvas || (this.rawDeltas && this.rawDeltas.length > 0);
+			},
+			needsFetch: function() {
+				return this.invalidFrom >= this.wireId || !this.hasContent();
 			}
 		};
 		this._emptyTilesCount += 1;
 		this._tiles[key] = tile;
+
+		if (this._debug)
+		{
+			tile._debugLoadTile = 0;
+			tile._debugLoadDelta = 0;
+			tile._debugInvalidateCount = 0;
+		}
+
 		return tile;
 	},
 
-	_tileHasContent: function(key) {
+	_tileNeedsFetch: function(key) {
 		var tile = this._tiles[key];
-		return tile && tile.hasContent();
+		return tile && tile.needsFetch();
```

Comment:
tiles: track invalidation by monotonic timestamp does seem to genuinely cause sheet_operation_spec.js 'Switching sheet sets the view that contains cell-cursor' to fail and it does seem to be the case that the cell cursor might be gone missing on switching sheets.

Problem seems to be here:  !tile || tile.needsFetch()

Reply:
Ah - so should be hasContent -> !needsFetch() I guess.